crates/syn2mas/src/synapse_reader/checks.rs

```diff
@@ -0,0 +1,290 @@
+    }
+
+    if synapse.server_name != mas_matrix.homeserver {
+        errors.push(CheckError::ServerNameMismatch {
+            synapse: synapse.server_name.clone(),
+            mas: mas_matrix.homeserver.clone(),
+        });
+    }
+
+    Ok((warnings, errors))
+}
+
+/// Check that the Synapse database is sane for migration. Returns a list of warnings and errors.
+///
+/// # Errors
+///
+/// - If there is some database connection error, or the given database is not a Synapse database.
+/// - If the OAuth2 section of the MAS configuration could not be parsed.
+#[tracing::instrument(skip_all)]
+pub async fn synapse_database_check(
+    synapse_connection: &mut PgConnection,
+    synapse: &Config,
+    mas: &Figment,
+) -> Result<(Vec<CheckWarning>, Vec<CheckError>), Error> {
+    #[derive(FromRow)]
+    struct UpstreamOAuthProvider {
+        auth_provider: String,
+        num_users: i64,
+    }
+
+    let mut errors = Vec::new();
+    let warnings = Vec::new();
+
+    let num_guests: i64 = query_scalar("SELECT COUNT(1) FROM users WHERE is_guest <> 0")
+        .fetch_one(&mut *synapse_connection)
+        .await?;
+    if num_guests > 0 {
+        errors.push(CheckError::GuestsInDatabase { num_guests });
+    }
+
+    let num_non_email_3pids: i64 =
+        query_scalar("SELECT COUNT(1) FROM user_threepids WHERE medium <> 'email'")
+            .fetch_one(&mut *synapse_connection)
+            .await?;
+    if num_non_email_3pids > 0 {
+        errors.push(CheckError::NonEmailThreepidsInDatabase {
+            num_non_email_3pids,
+        });
+    }
+
+    let oauth_provider_user_counts = query_as::<_, UpstreamOAuthProvider>("SELECT auth_provider, COUNT(*) AS num_users FROM user_external_ids GROUP BY auth_provider ORDER BY auth_provider").fetch_all(&mut *synapse_connection).await?;
+    if !oauth_provider_user_counts.is_empty() {
+        let syn_oauth2 = synapse.all_oidc_providers();
+        let mas_oauth2 = UpstreamOAuth2Config::extract_or_default(mas)?;
+        for row in oauth_provider_user_counts {
+            let matching_syn = syn_oauth2.iter().find(|syn_provider| {
+                let Some(ref idp_id) = syn_provider.idp_id else {
+                    return false;
+                };
+                &row.auth_provider == idp_id
+            });
+
+            let Some(matching_syn) = matching_syn else {
+                errors.push(CheckError::SynapseMissingOAuthProvider {
+                    provider: row.auth_provider,
+                    num_users: row.num_users,
+                });
+                continue;
+            };
+
+            // unwrap: we already check the issuer is present before accepting the config dict
+            let issuer = matching_syn.issuer.clone().unwrap();
+
+            // TODO extract this logic and reuse it for the migration
+            let matching_mas = mas_oauth2
+                .providers
+                .iter()
+                .find(|mas_provider| mas_provider.issuer == issuer);
```

Comment:
we've talked about the heuristic for matching the providers on both sides; I'll assume that when you implement that in the migration, you'll implement what we talked about (having a `synapse_idp_id` in the MAS config to match) and use it here as well

Reply:
fb8c3370ce0b169a948a07599bc9e97338686ea7